Adding State Updating and # of Tasks to LSTM and RGCN models

jzwart opened the pull request:
I added some functionality to the LSTM and RGCN models so that they can return the h and c states of the LSTM. The LSTM returns the most recent h and c states (i.e. most recent time step), while the RGCN returns the h and c states for every time step since it is looping through each timestep for the LSTM. 

I also added options for predicting either one or two tasks for the LSTM and RGCN. Previously, the models only predicted two tasks (i.e. you had to supply observations for two target features), but now both models should be able to predict either one or two tasks while defaulting to predicting only one tasks. This will be a big update and I didn't change `train.py` or other relevant functions that will call the LSTM/RGCN models because I think we need to first see if these proposed changes will work for all the projects. I also wasn't sure what to call the number of prediction tasks - I ended up calling them `tasks` since that is what @jsadler2 refers to them in his multi-task modeling paper, but I'm open to other suggestions (e.g. targets, target_features, etc..). I had to update some of the loss functions too since they defaulted to assuming the model was predicting 2 modeling tasks. 

Here's[ an example ](https://code.usgs.gov/wma/wp/run-pgdl-da/-/blob/training_model/test_river_dl_lstm.ipynb)of running these models with the new updates. This example is just predicting a single task and returning the h and c states. I show that both the LSTM and RGCN h and c states can be adjusted, updated, and influence the model predictions when supplied to the LSTM / RGCN as initial h and c states. This workflow also works for 2 tasks but it isn't shown right now. 

closes #98, #106

aappling-usgs commented on river_dl/rnns.py:
```diff
@@ -47,39 +69,52 @@ def train_step(self, data):
-            loss_aux = nnse_one_var_samplewise(y, y_pred, 1)
+            loss_main = nnse_one_var_samplewise(y, y_pred, 0, self.tasks)
+            if self.tasks == 2: 
+                loss_aux = nnse_one_var_samplewise(y, y_pred, 1, 2)
```
```suggestion
                loss_aux = nnse_one_var_samplewise(y, y_pred, 1, self.tasks)
```
even though we know that `self.tasks == 2` when we reach this line, it's still clearer to use the variable name

aappling-usgs commented on river_dl/RGCN.py:
```diff
@@ -108,25 +123,45 @@ def __init__(self, hidden_size, A, flow_in_temp=False, rand_seed=None):
+        h_list = []
+        c_list = []
         graph_size = self.A.shape[0]
         hidden_state_prev, cell_state_prev = (
             tf.zeros([graph_size, self.hidden_size]),
             tf.zeros([graph_size, self.hidden_size]),
         )
         out = []
         n_steps = inputs.shape[1]
+        h_update = tf.cast(kwargs['h_init'], tf.float32)
+        c_update = tf.cast(kwargs['c_init'], tf.float32)
+        if self.return_state:
+            # set the initial h & c states to the supplied h and c states if using DA 
+            hidden_state_prev = h_update 
+            cell_state_prev = c_update 
```
Is it necessary to define `h_update` and `c_update` here rather than waiting until line 184? They don't seem to get used before being overwritten, except in line 163 which itself doesn't seem to get used. (Or does `self.lstm` change the properties of `self`, such that line 163 does something even though `seq` and `state` go unused?) I'm wondering if this block of 12 lines could be simplified:
```suggestion
        out = []
        n_steps = inputs.shape[1]
        # set the initial h & c states to the supplied h and c states if using DA, or 0's otherwise
        if self.return_state:
            hidden_state_prev = tf.cast(kwargs['h_init'], tf.float32)
            cell_state_prev = tf.cast(kwargs['c_init'], tf.float32)
        else:
            hidden_state_prev = tf.zeros([graph_size, self.hidden_size])
            cell_state_prev = tf.zeros([graph_size, self.hidden_size])
```

jzwart replied:
good call, I think that still works and is much cleaner. I'm pretty sure the call on 163 was leftover from previous versions of the river-dl code as I had those lines in an earlier version.

aappling-usgs reviewed (commented):
Hi Jake, super cool to see these updates going into the shared river-dl codebase. I've made comments that reflect on decisions made by both you and Jeff, plus a few changes to Jeff's original code to maintain readability as the scope of this repo expands.

My one other major comment is about the interfaces you've designed here. I see this block in your ipynb:
```py
old_preds_lstm = model_lstm(inputs)
old_preds_rgcn = model_rgcn(inputs)
h_lstm, c_lstm = model_lstm.rnn_layer.states
h_rgcn = model_rgcn.h_gr
c_rgcn = model_rgcn.c_gr

updated_h_lstm = h_lstm * 500 # update just the states of just the last time step
updated_c_lstm = c_lstm * 500
updated_h_rgcn = h_rgcn[:,-1,:] * 500  # rgcn returns entire timeseries of h & c states so we need to select most recent states 
updated_c_rgcn = c_rgcn[:,-1,:] * 500 

# initialize the states with the previously trained states 
model_lstm.rnn_layer.reset_states(states=[updated_h_lstm, updated_c_lstm])
new_preds_lstm = model_lstm(inputs) 
new_preds_rgcn = model_rgcn(inputs, h_init = updated_h_rgcn, c_init = updated_c_rgcn) 
new_h_lstm, new_c_lstm = model_lstm.rnn_layer.states
new_h_rgcn = model_rgcn.h_gr
new_c_rgcn = model_rgcn.c_gr
```

I see these differences in interfaces:
1. To access or reset states in the LSTM, you accept or pass in a list, whereas to access or reset states in the RGCN, you accept or pass in two objects. 
2. RGCN returns the entire timeseries whereas LSTM returns just the most recent states; is there a need to preserve all values in the RGCN, or could that model be changed to store/change just the most recent state as well?
3. You reset states in LSTM by calling the built-in `reset_states` method. Could you write such a method for `RGCN` as well rather than adding arguments to `model_rgcn`?

Could you make it so that both models have the same interface? That would make it easier to switch between RCGN and LSTM in any code that uses the state information.

jsadler2 commented on river_dl/loss_functions.py:
```diff
@@ -99,23 +103,27 @@ def y_data_components(data, y_pred, var_idx):
+        rmse_main = rmse_masked_one_var(data, y_pred, 0, tasks)
+        if tasks == 2: 
+            rmse_aux = rmse_masked_one_var(data, y_pred, 1, tasks)
+            rmse_loss = rmse_main + lamb * rmse_aux
```
👍 . I think `lambda_aux` is a good idea.

jzwart replied:
updated to `lambda_aux` here and many other locations (`rnns.py`, `train.py`, `train_model.py`)

jsadler2 commented on river_dl/RGCN.py:
```diff
@@ -11,22 +11,36 @@
+        hidden_size, 
+        A, 
+        tasks=1, 
+        dropout=0,  # I propose changing this to 'recurrent_dropout' and adding another option for 'dropout' since these will map to the options for the tf LSTM layers https://www.tensorflow.org/api_docs/python/tf/keras/layers/LSTMCell ; and also https://arxiv.org/pdf/1512.05287.pdf 
```
I think this is a good idea to change the variable name.
```suggestion
        recurrent_dropout=0, 
```

jzwart replied:
updated

jsadler2 reviewed (commented):
@jzwart - like Alison said, great to see these additions to the shared codebase. I'm wondering if we should take what you've started and try to make it even more modular. 
1. **have the `RGCN` layer return the `h`'s and `c`'s** - I think this makes a lot of sense. As it is written now, the RGCN and the output "layers" are both in one layer class. If we just return the `h`'s and `c`'s we can swap out the output layer without changing the actual `RGCN` code. This will be useful for if we want to expand to say 3 outputs later on or if we want the model to predict the probability distribution.
2. **multi-task loss functions** - I think we should rename and rewrite the loss functions so it is clear they are _multi-task_ loss functions. These functions will take the a 1-d array of lambdas, one for each output variable. That way we don't have to explicitly have a different function for 1, 2, 3, ... variables. 
3. **separate `SingleTaskLSTMModel`** - I think it might make sense to have a separate `SingleTaskLSTMModel` class. It would be a very simple class, but I think it could make the code more maintainable/clear because we wouldn't have to write in all this logic about `lambda`, `num_tasks`, etc. That logic I guess would be upstream. If someone was running just a single-task model, they'd just use the `SingleTaskLSTMModel` and not the other one (maybe we call that one a `MultiTaskLSTMModel`).
4. (something to think about) **One multi-task output layer** - In line with simplifying, I wonder if we could just have all of the "tasks" share just one output layer. This would really simplify the model definition. We could just have pass (or infer) the number of outputs and then the output layer would output that number. That would preclude the fancy gradient correction routine, but I'm not sure we need that (at least in the main branch).

If you are good with it, Jake, I can commit code to your branch that addresses 1, 2, and 3 above.

jsadler2 commented on river_dl/rnns.py:
```diff
@@ -4,38 +4,60 @@
-        aux_prediction = self.dense_aux(x)
-        return tf.concat([main_prediction, aux_prediction], axis=2)
+        if self.return_state: 
+            x, h, c = self.rnn_layer(inputs)
```
does this make the `h` and `c` accessible? i see they are returned, but i don't see how they can be accessed from the `LSTMModel` object.

jzwart replied:
I've accessed `h` and `c` through the already existing `rnn_layer.states` call, so I don't think they're accessed via the `LSTMModel` object like they are in the `RGCNModel`. @aappling-usgs has a comment above for making the models access and reset the states more similarly and I think addressing that comment will fix how the states are accessed here.

jzwart commented on river_dl/RGCN.py:
```diff
@@ -11,22 +11,36 @@
+        hidden_size, 
+        A, 
+        tasks=1, 
+        dropout=0,  # I propose changing this to 'recurrent_dropout' and adding another option for 'dropout' since these will map to the options for the tf LSTM layers https://www.tensorflow.org/api_docs/python/tf/keras/layers/LSTMCell ; and also https://arxiv.org/pdf/1512.05287.pdf 
```
updated

jzwart commented:
> Could you make it so that both models have the same interface? That would make it easier to switch between RCGN and LSTM in any code that uses the state information.

Yes, I think that's a good idea and will update so that they have similar interfaces. 

> If you are good with it, Jake, I can commit code to your branch that addresses 1, 2, and 3 above. 

That'd be great @jsadler2  , I will add a few other commits based on @aappling-usgs  and your comments as well.

jzwart commented on river_dl/rnns.py:
```diff
@@ -4,38 +4,60 @@
         self.rnn_layer = layers.LSTM(
             hidden_size,
             return_sequences=True,
+            stateful=True,
```
Good point. I don't think the non-DA applications would need to access the states, unless they want to see how different states relate to input or something like that. Maybe this could be set to the same as `return_state` - so it will be `True` when states are needed, otherwise, `False`

jzwart commented on river_dl/RGCN.py:
```diff
@@ -108,25 +123,45 @@ def __init__(self, hidden_size, A, flow_in_temp=False, rand_seed=None):
+            cell_state_prev = c_update 
         for t in range(n_steps):
+            seq, state = self.lstm(inputs[:, t, :], states=[h_update, c_update])
+            h, c = state # are these used anywhere? 
```
removed

jzwart commented on river_dl/RGCN.py:
```diff
@@ -108,25 +123,45 @@ def __init__(self, hidden_size, A, flow_in_temp=False, rand_seed=None):
         )
         out = []
         n_steps = inputs.shape[1]
+        h_update = tf.cast(kwargs['h_init'], tf.float32)
```
yup, Alison added a suggestion that should fix this.

jzwart commented:
OK @jsadler2 , I've addressed most but not all comments. If you want to commit code to this branch addressing your points 1, 2, and 3 above, then go ahead. I won't commit code for a bit.

jsadler2 commented on river_dl/RGCN.py:
```diff
@@ -11,22 +11,36 @@
         :param A: [numpy array] adjacency matrix
+        :param tasks: [int] number of prediction tasks to perform - currently supports either 1 or 2 prediction tasks 
+        :param dropout: [float] value between 0 and 1 for the probability of a reccurent element to be zero  
         :param flow_in_temp: [bool] whether the flow predictions should feed
```
I think this should be moved to the `model` level, not the `layer` level because we will be separating out the output layer from the RGCN layer.

jsadler2 commented:
Okay @jzwart - I added a few commits that get at my 1, 2, and 3 points from above. I'm realizing that there is some more simplifying that I think can/should be done here, but I'm not sure if it belongs in this PR. For example, taking out the sample weights relates to the loss functions (#98), but that may be beyond scope here.

jzwart commented on river_dl/rnns.py:
```diff
@@ -5,34 +5,77 @@
+        self.rnn_layer = layers.LSTM(
+            hidden_size,
+            return_sequences=True,
+            stateful=True,
+            return_state=True,
```
Should `return_state` be input to the model call? i.e. do we always want the states to be returned?

jsadler2 replied:
Good question. To simplify, I took out the argument. I did this in the other models too. I figured that we can assign them to the `self.h` and `self.c` attributes always and then folks can decide whether or not to actually go in and access them. But they won't be returned in the `call` function, so I don't think it will gum anything up. Does that seem like a good idea? I'm fine going back to having it as an argument too.

aappling-usgs replied:
It seems smart to have the format of outputs stay the same regardless of what argument values are passed in.

jzwart commented on river_dl/RGCN.py:
```diff
@@ -157,42 +138,72 @@ def call(self, inputs, **kwargs):
+        h_init = kwargs.get('h_init', tf.zeros([batch_size, self.hidden_size]))
+        c_init = kwargs.get('c_init', tf.zeros([batch_size, self.hidden_size]))
+        h_gr, c_gr = self.rgcn_layer(inputs, h_init=h_init, c_init=c_init)
+        self.h_gr = h_gr
+        self.c_gr = c_gr
```
could make these the last timestep in the h & c list to make more similar to the LSTM model (which only returns the most recent h & c states). 
e.g. `self.h_gr = h_gr[:,-1,:]`

jsadler2 replied:
Yeah. We could do that. I'm fine either way.

jzwart replied:
I think I'd prefer the h and c states returned for both RGCN and LSTM to be the same length

jsadler2 replied:
Sounds good to me. That makes sense.

jzwart commented:
looks good @jsadler2 , I added a couple comments

jsadler2 commented on river_dl/RGCN.py:
```diff
@@ -157,42 +138,72 @@ def call(self, inputs, **kwargs):
+        h_init = kwargs.get('h_init', tf.zeros([batch_size, self.hidden_size]))
+        c_init = kwargs.get('c_init', tf.zeros([batch_size, self.hidden_size]))
+        h_gr, c_gr = self.rgcn_layer(inputs, h_init=h_init, c_init=c_init)
+        self.h_gr = h_gr
+        self.c_gr = c_gr
```
```suggestion
        self.h_gr = h_gr[:, -1, :]
        self.c_gr = c_gr[:, -1, :]
```

jsadler2 commented on river_dl/RGCN.py:
```diff
@@ -176,23 +211,77 @@ def call(self, inputs, **kwargs):
+            return_state)
+            
+        self.h_gr = None
+        self.c_gr = None
```
What do you think about this idea, @jzwart? If we do this, you'd access it like 

```python
model = RGCNModel(bla, bla, bla)
h, c = model.states # would just be the final states
```

jzwart replied:
Yeah I think that would be good

jsadler2 commented on river_dl/predict.py:
```diff
@@ -320,7 +316,8 @@ def predict_from_arbitrary_data(
     model = load_model_from_weights(
-        model_type, model_weights_dir, hidden_size, dist_matrix, flow_in_temp,
+        model_type, model_weights_dir, hidden_size, dist_matrix,
+        um_tasks=num_tasks
```
```suggestion
        num_tasks=num_tasks
```

jsadler2 commented:
@jzwart  - I got a little carried away :). I felt like just one thing led to another and I ended doing a couple more major changes that related to what you started.

So now in addition to that you wrote in, i.e.,:  
* the `states` attribute in both the `RGCNModel` and the `LSTMModel` being exposed 
* the `num_tasks` argument 


I also 
* removed the use of weights in the loss functions and took that out of the `train` script (#98)
* unified the multitasking approach so that now the `LSTMModel` and the `GRUModel` are much simpler (#106)

These two additional changes were mostly prompted by the `num_tasks` option that you introduced.

I think all the changes together will make the codebase both simpler and more flexible.

I'm done making changes, now, so if you could look over what I've done, that'd be great.

jzwart commented on river_dl/train_model.py:
```diff
@@ -45,31 +55,56 @@
 parser.add_argument(
-    "--lamb", help="lambda for weighting aux gradient", default=1.0, type=float
+    "--num_tasks",
+    help="number of tasks (outputs to be predicted)",
```
```suggestion
    help="number of tasks (variables to be predicted)",
```

jzwart commented:
I think it looks good @jsadler2 ! I like the unified approach to all the models and similar output for predictions and states.  There is a slight difference in how the LSTM and RGCN states are reset and I wonder if we could make them the same. For the RGCN we'd reset by supplying the `h` and `c` states: 
```
rgcn_preds = rgcn_model(inputs = inputs, h_init = h, c_init = c) 
```
but the LSTM states would need to be reset before making predictions rather than supplying to the function: 
```
lstm_model.rnn_layer.reset_states(states = [h, c]) 
lstm_preds = lstm_model(inputs = inputs) 
```
It'd be nice to make those the same so we don't have to change the prediction / training code as much depending on the model type we're using. I _think_ we'd just need to add an `h_init` and `c_init` argument to the LSTMModel call and reset the states in there (either to zero or the supplied `h` and `c` states). I'll add a suggestion where I think it would go. 

Other than that, I think the changes look good and it can be merged

jzwart commented on river_dl/rnns.py:
```diff
@@ -2,126 +2,68 @@
-            # adjust auxiliary gradient
-            gradient_shared_aux = adjust_gradient_list(
-                gradient_shared_main, gradient_shared_aux, self.grad_log_file
+        x, h, c = self.rnn_layer(inputs)
```
I propose adding in the following right above this line to make the reset of the `h` and `c` states more similar to RGCN model: 
```
batch_size = inputs.shape[0]
h_init = kwargs.get("h_init", tf.zeros([batch_size, self.hidden_size]))
c_init = kwargs.get("c_init", tf.zeros([batch_size, self.hidden_size]))
self.rnn_layer.reset_states(states = [h_init, c_init]) 
```
I _think_ this should work?

jsadler2 replied:
good idea

jzwart commented:
great, looks good to me. Merging
